File: app/src/androidTest/java/pavelnazimok/uitestingfeatures/java/tests/BaseTest.java
package pavelnazimok.uitestingfeatures.java.tests;

import android.Manifest;

import androidx.test.rule.ActivityTestRule;
import androidx.test.rule.GrantPermissionRule;

import org.junit.Rule;
import org.junit.rules.ExternalResource;
import org.junit.rules.RuleChain;
import org.junit.runner.Description;
import org.junit.runners.model.Statement;

import pavelnazimok.uitestingfeatures.MainActivity;
import pavelnazimok.uitestingfeatures.java.robots.DeviceRobot;
import pavelnazimok.uitestingfeatures.java.robots.FirstScreenRobot;
import pavelnazimok.uitestingfeatures.java.robots.MenuRobot;
import pavelnazimok.uitestingfeatures.java.robots.SecondScreenRobot;
import pavelnazimok.uitestingfeatures.java.robots.ThirdScreenRobot;
import pavelnazimok.uitestingfeatures.java.utils.CustomAnnotation;

import static pavelnazimok.uitestingfeatures.java.utils.AndroidExtensions.isNetworkConnected;

public abstract class BaseTest {

    final FirstScreenRobot firstScreenRobot = new FirstScreenRobot();
    final SecondScreenRobot secondScreenRobot = new SecondScreenRobot();
    final ThirdScreenRobot thirdScreenRobot = new ThirdScreenRobot();
    final MenuRobot menuRobot = new MenuRobot();
    final DeviceRobot deviceRobot = new DeviceRobot();

    private GrantPermissionRule runtimePermissionRule = GrantPermissionRule.grant(
            Manifest.permission.ACCESS_COARSE_LOCATION
    );

    private ExternalResource setUpRule = new ExternalResource() {
        @Override
        public Statement apply(final Statement base, final Description description) {
            if (description.getAnnotation(CustomAnnotation.class) != null) {
                System.out.println("This test is annotated with CustomAnnotation");
            }

            return super.apply(base, description);
        }

        @Override
        protected void before() {
            if (!isNetworkConnected()) {
                deviceRobot.setNetworkEnabled(true);
            }

            activityTestRule.launchActivity(null);

            deviceRobot.wakeUp();
        }
    };

    private ExternalResource tearDownRule = new ExternalResource() {
        @Override
        protected void after() {
            deviceRobot.enableNetworkAfterTest();
        }
    };

    private ActivityTestRule<MainActivity> activityTestRule = new ActivityTestRule<>(MainActivity.class, true, false);

    @Rule
    public RuleChain ruleChain = RuleChain.emptyRuleChain()
            .around(runtimePermissionRule)
            .around(setUpRule)
            .around(activityTestRule)
            .around(tearDownRule);
}

File: app/src/androidTest/java/pavelnazimok/uitestingfeatures/kotlin/tests/BaseTest.kt
package pavelnazimok.uitestingfeatures.kotlin.tests

import android.Manifest
import android.content.Intent
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.rule.ActivityTestRule
import androidx.test.rule.GrantPermissionRule
import org.junit.Rule
import org.junit.rules.ExternalResource
import org.junit.rules.RuleChain
import org.junit.runner.Description
import org.junit.runner.RunWith
import org.junit.runners.model.Statement
import pavelnazimok.uitestingfeatures.MainActivity
import pavelnazimok.uitestingfeatures.kotlin.robots.NETWORK_WAS_DISABLED_DURING_TEST
import pavelnazimok.uitestingfeatures.kotlin.robots.onDevice
import pavelnazimok.uitestingfeatures.kotlin.utils.CustomAnnotation
import pavelnazimok.uitestingfeatures.kotlin.utils.isNetworkConnected
import pavelnazimok.uitestingfeatures.kotlin.utils.targetContext

@RunWith(AndroidJUnit4::class)
abstract class BaseTest {

    private val runtimePermissionRule = GrantPermissionRule.grant(
            Manifest.permission.ACCESS_COARSE_LOCATION
    )

    private val setUpRule = object : ExternalResource() {
        override fun apply(base: Statement, description: Description): Statement {
            if (description.getAnnotation(CustomAnnotation::class.java) != null) {
                println("This test is annotated with CustomAnnotation")
            }

            return super.apply(base, description)
        }

        override fun before() {
            if (!isNetworkConnected) {
                onDevice {
                    setNetworkEnabled(true)
                }
            }

            activityTestRule.launchActivity(null)
        }
    }

    private val tearDownRule = object : ExternalResource() {
        override fun after() {
            if (NETWORK_WAS_DISABLED_DURING_TEST) {
                onDevice {
                    setNetworkEnabled(true)
                }
            }
        }
    }

    private val activityTestRule = ActivityTestRule(MainActivity::class.java, true, false)

    @get:Rule
    var ruleChain: RuleChain = RuleChain.emptyRuleChain()
            .around(runtimePermissionRule)
            .around(setUpRule)
            .around(activityTestRule)
            .around(tearDownRule)

    internal fun openMainActivity() {
        val mainActivityIntent = Intent(targetContext, MainActivity::class.java)
        mainActivityIntent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
        targetContext.startActivity(mainActivityIntent)
    }
}